File: 섹션07. BFS - 2차원 배열 탐색/06. BOJ 7562 - 나이트의 이동.cpp
#include <stdio.h>

#define MAX (300 + 30)

int T, L;
int sr, sc, er, ec;
int MAP[MAX][MAX];

struct RC
{
	int r;
	int c;
};

RC queue[MAX * MAX];

int dr[] = { -2, -1, 1, 2, 2, 1, -1, -2 };
int dc[] = { 1, 2, 2, 1, -1, -2, -2, -1 };

void input()
{
	scanf("%d %d %d %d %d", &L, &sr, &sc, &er, &ec);

	for (int r = 0; r < L; r++)
		for (int c = 0; c < L; c++)
			MAP[r][c] = 0;
}

void printMap() // for debug
{
	for (int r = 0; r < L; r++)
	{
		for (int c = 0; c < L; c++)
			printf("%d ", MAP[r][c]);
		putchar('\n');
	}
	putchar('\n');
}

int BFS(int r, int c)
{
	int rp, wp;

	rp = wp = 0;

	queue[wp].r = r;
	queue[wp++].c = c;

	MAP[r][c] = 1;

	while (rp < wp)
	{
		RC out = queue[rp++];

		if (out.r == er && out.c == ec)
			return MAP[out.r][out.c] - 1;

		for (int i = 0; i < 8; i++)
		{
			int nr, nc;

			nr = out.r + dr[i];
			nc = out.c + dc[i];

			if (nr < 0 || nc < 0 || nr > L - 1 || nc > L - 1) continue;
			if (MAP[nr][nc] != 0) continue;

			queue[wp].r = nr;
			queue[wp++].c = nc;

			MAP[nr][nc] = MAP[out.r][out.c] + 1;
		}
	}

	return -1; // error
}

int main()
{
	scanf("%d", &T);

	for (int tc = 0; tc < T; tc++)
	{
		input();

		printf("%d\n", BFS(sr, sc));
	}

	return 0;
}

File: Python/섹션07. BFS - 2차원 배열 탐색/06. BOJ 7562 - 나이트의 이동.py
MAX = 300 + 30

MAP = [[0] * MAX for _ in range(MAX)]

class RC:
    def __init__(self, r: int, c: int):
        self.r = r
        self.c = c
    
queue = [RC(0, 0) for _ in range(MAX * MAX)]   

dr = [-2, -1, 1, 2, 2, 1, -1, -2]
dc = [1, 2, 2, 1, -1, -2, -2, -1]

def input_data():             
    global L, sr, sc, er, ec, MAP
    
    L = int(input())
    sr, sc = map(int, input().split())
    er, ec = map(int, input().split())
    
    MAP = [[0] * MAX for _ in range(MAX)]
    
def print_map(): # for debug
    for r in range(L):
        print(*MAP[r][:L])
    print("")

def BFS(r, c):
    rp, wp = 0, 0
    
    queue[wp].r = r
    queue[wp].c = c
    wp += 1    
                    
    MAP[r][c] = 1                    
                    
    while rp < wp:
        out = queue[rp]
        rp += 1
        
        if out.r == er and out.c == ec:
            return MAP[out.r][out.c] - 1
                
        for i in range(8):
            nr = out.r + dr[i]
            nc = out.c + dc[i]
            
            if nr < 0 or nc < 0 or nr > L - 1 or nc > L - 1: continue
            if MAP[nr][nc] != 0: continue
           
            queue[wp].r = nr
            queue[wp].c = nc
            wp += 1            
            
            MAP[nr][nc] = MAP[out.r][out.c] + 1

    return -1 # error
               
T = int(input())
for _ in range(T):
    input_data()                 
                                      
    print(BFS(sr, sc))